File: backend/src/main/java/kr/co/darumtech/intra/domain/schedule/Schedule.java
package kr.co.darumtech.intra.domain.schedule;

import jakarta.persistence.*;
import kr.co.darumtech.intra.domain.customer.Customer;
import kr.co.darumtech.intra.domain.customer.Product;
import kr.co.darumtech.intra.domain.customer.Support;
import kr.co.darumtech.intra.domain.employee.Employee;
import lombok.*;

import java.time.LocalDate;
import java.time.LocalTime;

@Entity
@Table(name = "SCHEDULE")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@ToString
public class Schedule {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "NO")
    private Long no;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "EMPNO", nullable = false)
    private Employee employee;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "CUSTNO")
    private Customer customer;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "PRODNO")
    private Product product;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "SUPPNO")
    private Support support;

    @Column(name = "CONTENTS", length = 500)
    private String contents;

    @Column(name = "LOCATION", length = 200)
    private String location;

    @Column(name = "WORK_DATE", nullable = false)
    private LocalDate workDate;

    @Column(name = "STIME")
    private LocalTime stime;

    @Column(name = "RTIME")
    private LocalTime rtime;

    @Column(name = "ETIME")
    private LocalTime etime;

    @Column(name = "START_NO", nullable = false)
    private long startNo = 0L;

    @Column(name = "HOLIDAY")
    private Boolean holiday = false;

    @PrePersist
    @PreUpdate
    private void ensureDefaults() {
        // ensure non-null start number for DB constraint
        if (startNo == 0L && no != null) {
            startNo = no;
        }
        if (holiday == null) {
            holiday = false;
        }
    }
}

File: frontend/src/types/schedule.ts
/**
 * 스케줄 관련 TypeScript 타입 정의
 */

/**
 * 스케줄 DTO
 */
export interface Schedule {
  no: number;
  empno: number;
  deptno?: number;
  employeeName: string;
  departmentName?: string;
  departmentRank?: number;
  jobgradeName?: string;
  custno?: number;
  customerName?: string;
  prodno?: number;
  productName?: string;
  suppno?: number;
  supportName?: string;
  contents?: string;
  location?: string;
  workDate: string; // ISO date format (YYYY-MM-DD)
  stime?: string; // ISO time format (HH:mm:ss)
  rtime?: string;
  etime?: string;
  startNo?: number;
  holiday: boolean;
}

/**
 * 스케줄 생성 요청
 */
export interface CreateScheduleRequest {
  empno: number;
  custno?: number;
  prodno?: number;
  suppno?: number;
  contents?: string;
  location?: string;
  /**
   * 시작 날짜 (범위 방식 사용 시)
   * dates 필드가 있으면 무시됨
   */
  startDate?: string;
  /**
   * 종료 날짜 (범위 방식 사용 시)
   * dates 필드가 있으면 무시됨
   */
  endDate?: string;
  /**
   * 개별 날짜 목록 (직접 선택 방식)
   * 이 필드가 있으면 startDate/endDate보다 우선 처리됨
   */
  dates?: string[];
  stime?: string;
  rtime?: string;
  etime?: string;
  startNo?: number;
  holiday?: boolean;
}

/**
 * 스케줄 수정 요청
 */
export interface UpdateScheduleRequest {
  custno?: number;
  prodno?: number;
  suppno?: number;
  contents?: string;
  location?: string;
  startDate?: string;
  endDate?: string;
  stime?: string;
  rtime?: string;
  etime?: string;
  startNo?: number;
  holiday?: boolean;
}

/**
 * 스케줄 필터 조건
 */
export interface ScheduleFilter {
  empno?: number;
  deptno?: number;
  custno?: number;
  startDate: string;
  endDate: string;
  holiday?: boolean;
}
